applications/TrilinosApplication/trilinos_space.h

```diff
@@ -8,36 +8,37 @@
 //					 Kratos default license: kratos/license.txt
 //
 //  Main authors:    Riccardo Rossi
+//  Collaborator:    Vicente Mataix Ferrandiz
 //
 
-#if !defined(KRATOS_TRILINOS_SPACE_H_INCLUDED )
-#define  KRATOS_TRILINOS_SPACE_H_INCLUDED
+#pragma once
 
 // System includes
 
 // External includes
 
-// Trilinos includes
-#include "Epetra_Import.h"
-#include "Epetra_MpiComm.h"
-#include "Epetra_Map.h"
-#include "Epetra_Vector.h"
-#include "Epetra_FEVector.h"
-#include "Epetra_FECrsMatrix.h"
-#include "Epetra_IntSerialDenseVector.h"
-#include "Epetra_SerialDenseMatrix.h"
-#include "Epetra_SerialDenseVector.h"
-#include "EpetraExt_CrsMatrixIn.h"
+/* Trilinos includes */
+#include <Epetra_Import.h>
+#include <Epetra_MpiComm.h>
+#include <Epetra_Map.h>
+#include <Epetra_Vector.h>
+#include <Epetra_FEVector.h>
+#include <Epetra_FECrsMatrix.h>
+#include <Epetra_IntSerialDenseVector.h>
+#include <Epetra_SerialDenseMatrix.h>
+#include <Epetra_SerialDenseVector.h>
+#include <EpetraExt_CrsMatrixIn.h>
 #include <EpetraExt_VectorIn.h>
 #include <EpetraExt_RowMatrixOut.h>
 #include <EpetraExt_MultiVectorOut.h>
 
 // Project includes
-#include "includes/define.h"
 #include "includes/ublas_interface.h"
+#include "spaces/ublas_space.h"
+#include "includes/parallel_environment.h"
```

Comment:
Both are required:

~~~sh
In file included from /home/opc/src/Kratos/applications/TrilinosApplication/external_includes/amgcl_mpi_solver_impl.cpp:39:
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h: At global scope:
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:586:15: error: ‘SCALING_DIAGONAL’ does not name a type
         const SCALING_DIAGONAL ScalingDiagonal = SCALING_DIAGONAL::NO_SCALING,
               ^~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:640:15: error: ‘SCALING_DIAGONAL’ does not name a type
         const SCALING_DIAGONAL ScalingDiagonal = SCALING_DIAGONAL::NO_SCALING,
               ^~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:586:50: error: ‘SCALING_DIAGONAL’ has not been declared
         const SCALING_DIAGONAL ScalingDiagonal = SCALING_DIAGONAL::NO_SCALING,
                                                  ^~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:587:53: error: ‘ParallelEnvironment’ has not been declared
         const DataCommunicator& rDataCommunicator = ParallelEnvironment::GetDefaultDataCommunicator()
                                                     ^~~~~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:640:50: error: ‘SCALING_DIAGONAL’ has not been declared
         const SCALING_DIAGONAL ScalingDiagonal = SCALING_DIAGONAL::NO_SCALING,
                                                  ^~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:641:53: error: ‘ParallelEnvironment’ has not been declared
         const DataCommunicator& rDataCommunicator = ParallelEnvironment::GetDefaultDataCommunicator()
                                                     ^~~~~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:667:53: error: ‘ParallelEnvironment’ has not been declared
         const DataCommunicator& rDataCommunicator = ParallelEnvironment::GetDefaultDataCommunicator()
                                                     ^~~~~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:691:53: error: ‘ParallelEnvironment’ has not been declared
         const DataCommunicator& rDataCommunicator = ParallelEnvironment::GetDefaultDataCommunicator()
                                                     ^~~~~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:705:53: error: ‘ParallelEnvironment’ has not been declared
         const DataCommunicator& rDataCommunicator = ParallelEnvironment::GetDefaultDataCommunicator()
                                                     ^~~~~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:728:53: error: ‘ParallelEnvironment’ has not been declared
         const DataCommunicator& rDataCommunicator = ParallelEnvironment::GetDefaultDataCommunicator()
                                                     ^~~~~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h: In static member function ‘static double Kratos::TrilinosSpace<TMatrixType, TVectorType>::GetScaleNorm(const Kratos::ProcessInfo&, const MatrixType&, int, const Kratos::DataCommunicator&)’:
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:645:18: error: ‘SCALING_DIAGONAL’ has not been declared
             case SCALING_DIAGONAL::NO_SCALING:
                  ^~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:647:18: error: ‘SCALING_DIAGONAL’ has not been declared
             case SCALING_DIAGONAL::CONSIDER_PRESCRIBED_DIAGONAL: {
                  ^~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:651:18: error: ‘SCALING_DIAGONAL’ has not been declared
             case SCALING_DIAGONAL::CONSIDER_NORM_DIAGONAL:
                  ^~~~~~~~~~~~~~~~
/home/opc/src/Kratos/applications/TrilinosApplication/trilinos_space.h:653:18: error: ‘SCALING_DIAGONAL’ has not been declared
             case SCALING_DIAGONAL::CONSIDER_MAX_DIAGONAL:
                  ^~~~~~~~~~~~~~~~
~~~

Reply:
ah now I see

The DefaultDataComm must not be used here, please remove the defaults

Otherwise it is very dangerous because using a MPI_Comm that is not world can easily leads to deadlocks if the default is used somewhere deep in the code

Reply:
Okidoki